### PR TITLE
SPRIN-10: Product Type Filter

### PR DESCRIPTION
**Pull request by m-cherniak**

### CODE REVIEW

File: src/main/java/fragments/FilterFragment.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import common.BasePage;
+import org.openqa.selenium.WebDriver;
+
+public class FilterFragment extends BasePage {
+
+    public FilterFragment(WebDriver driver) {
+        super(driver);
+    }
+
+    private static final String SHOES_CATEGORY = "//div[@class='filter-options']//span[normalize-space(text())='Shoes']";
```
Comment (vitalik1985o) [inline src/main/java/fragments/FilterFragment.java:12]: Make locator universal
```suggestion
    private static final String SHOES_CATEGORY = "//div[@class='filter-options']//span[normalize-space(text())='%s']";
```

File: src/test/java/tests/ProductTypeFilter.java
Comment (Goldbe54) [inline src/test/java/tests/ProductTypeFilter.java:1]: "Select a filter by broduct type" step is not checked and reprodused. Add the necessary steps and checks

**Comment by m-cherniak** [comment]
https://vitaliyosyp9282.atlassian.net/jira/software/projects/SPRIN/boards/3?selectedIssue=SPRIN-10

File: src/main/java/fragments/FilterFragment.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    protected void waitUntilElementInvisible(By locator) {
+        try {
+            wait.until(ExpectedConditions.invisibilityOfElementLocated(locator));
+        } catch (WebDriverException e) {
+            logger.warn("Element is still visible: " + locator);
+        }
+    }
+    public void waitUntilSpinnerDisappear() {
```
Comment (mykolazubchyk) [inline src/main/java/fragments/FilterFragment.java:34]: add a line between 33 and 34

File: src/main/java/pages/WomanPage.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package pages;
+
+import common.BasePage;
+import fragments.FilterFragment;
+import fragments.HeaderFragment;
+import fragments.ItemCatalogFragment;
+import org.openqa.selenium.WebDriver;
+
+public class WomanPage extends BasePage {
+    private FilterFragment filterFragment;
```
Comment (mykolazubchyk) [inline src/main/java/pages/WomanPage.java:10]: add a line between 9 and 10 lines

File: src/main/java/pages/WomanPage.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    private ItemCatalogFragment itemCatalogFragment;
+    private HeaderFragment headerFragment;
+
+    public WomanPage (WebDriver driver){
+        super(driver);
+        this.filterFragment = new FilterFragment(driver);
```
Comment (mykolazubchyk) [inline src/main/java/pages/WomanPage.java:16]: And here too

File: src/test/java/tests/ProductTypeFilter.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import pages.WomanPage;
+
+public class ProductTypeFilter extends BaseTest {
+
+    @Test(description = "SPRIN-10:Product Type Filter ")
+    public void verifyProductTypeFilter() throws InterruptedException {
+        WomanPage womanPage = new WomanPage(driver);
```
Comment (mykolazubchyk) [inline src/test/java/tests/ProductTypeFilter.java:11]: Move object to before method

File: src/test/java/tests/ProductTypeFilter.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+        womanPage.getHeaderFragment().acceptCookies();
```
Comment (mykolazubchyk) [inline src/test/java/tests/ProductTypeFilter.java:13]: move headerFragment.acceptCookies(); to BeforeMethod

**Review by mykolazubchyk — APPROVED** [review]

File: src/test/java/tests/ProductTypeFilter.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    public void verifyProductTypeFilter() {
+        WomanPage womanPage = new WomanPage(driver);
+
+        womanPage.getHeaderFragment().chooseCategoryDropdown("Women");
+        womanPage.getFilterFragment().waitUntilSpinnerDisappear();
+        womanPage.getFilterFragment().chooseCategory("Shoes");
+        womanPage.getFilterFragment().waitUntilSpinnerDisappear();
+
+        for (String name : womanPage.getItemCatalogFragment().getListOfAllItemsName()) {
+            Assert.assertTrue(name.contains("Shoes"), "Item name does not contain chosen category");
```
Comment (vitalik1985o) [inline src/test/java/tests/ProductTypeFilter.java:19]: use SoftAssert

File: src/main/java/fragments/FilterFragment.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        waitClickabilityOfElement(String.format(PRODUCT_TYPE,productType)).click();
+    }
+
+    protected void waitUntilElementInvisible(By locator) {
+        try {
+            wait.until(ExpectedConditions.invisibilityOfElementLocated(locator));
+        } catch (WebDriverException e) {
+            logger.warn("Element is still visible: " + locator);
+        }
+    }
```
Comment (Goldbe54) [inline src/main/java/fragments/FilterFragment.java:33]: Move it to `CommonActions` class

```suggestion
    protected void waitUntilElementDisappear(String locator) {
        try {
            wait.until(ExpectedConditions.invisibilityOfElementLocated(By.xpath(locator))));
        } catch (WebDriverException e) {
            logger.warn("Element is still visible: " + locator);
        }
    }
```

File: src/main/java/fragments/ItemCatalogFragment.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    private static final String LIST_OF_ITEM_NAMES = "//span[@class='product-block-name-wrapper']";
+
+    public ItemCatalogFragment(WebDriver driver) {
+        super(driver);
+    }
+
+    public List<String> getListOfAllItemsName() {
+        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.xpath(LOADING_SPINNER)));
+        return getListOfString(waitVisibilityOfElements(LIST_OF_ITEM_NAMES));
+    }
```
Comment (Goldbe54) [inline src/main/java/fragments/ItemCatalogFragment.java:24]:
```suggestion
    public List<String> getListOfAllItemsName() {
        return getListOfString(waitVisibilityOfElements(LIST_OF_ITEM_NAMES));
    }
```

File: src/main/java/pages/WomanPage.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public ItemCatalogFragment getItemCatalogFragment() {
+        return itemCatalogFragment;
+    }
+
+    public HeaderFragment getHeaderFragment() {
+        return headerFragment;
+    }
+}
```
Comment (Goldbe54) [inline src/main/java/pages/WomanPage.java:35]:
```suggestion
public class WomanPage extends BasePage {

    @Getter
    private final FilterFragment filterFragment = new FilterFragment(driver);
    @Getter
    private final ItemCatalogFragment itemCatalogFragment = new ItemCatalogFragment(driver);
    @Getter
    private final HeaderFragment headerFragment = new HeaderFragment(driver);

    public WomanPage(WebDriver driver) {
        super(driver);
    }
}
```

File: src/test/java/tests/ProductTypeFilter.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        womanPage.getFilterFragment().chooseProductType("Slip-ons");
+
+        for (String name : womanPage.getItemCatalogFragment().getListOfAllItemsName()) {
```
Comment (Goldbe54) [inline src/test/java/tests/ProductTypeFilter.java:24]: add speaner wait before get list

**Comment by m-cherniak** [comment]
image
https://vitaliyosyp9282.atlassian.net/jira/software/projects/SPRIN/boards/3?selectedIssue=SPRIN-10
@Goldbe54

File: src/main/java/common/BasePage.java [hunk 12]
```diff
@@ -33,6 +34,7 @@ protected WebElement waitVisibilityOfElement(String locator) {
         }
     }
 
+
```
Comment (Goldbe54) [inline src/main/java/common/BasePage.java:37]: delete empty line

File: src/main/java/common/CommonActions.java [hunk 13]
```diff
@@ -14,10 +14,9 @@
 import java.util.List;
 import java.util.Random;
 
-import static org.openqa.selenium.Keys.*;
-
 public class CommonActions extends BasePage {
 
+
```
Comment (Goldbe54) [inline src/main/java/common/CommonActions.java:19]: delete empty line

File: src/main/java/common/CommonActions.java [hunk 14]
```diff
@@ -85,7 +84,7 @@ protected static void sleep(long milliseconds) {
     @Attachment(value = "Page screenshot", type = "image/png")
     public static byte[] captureScreen(WebDriver driver) {
         File screenshot = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
-        FileUtils.copyFile(screenshot, new File("./target/allure-results//" + screenshot.getName()));
+        FileUtils.copyFile(screenshot, new File("./target/allure-results/screenshots//" + screenshot.getName()));
```
Comment (Goldbe54) [inline src/main/java/common/CommonActions.java:87]:
```suggestion
        FileUtils.copyFile(screenshot, new File("./target/allure-results//" + screenshot.getName()));
```

File: src/main/java/fragments/FilterFragment.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
+package fragments;
+
+import common.BasePage;
+import common.CommonActions;
+import org.openqa.selenium.WebDriver;
+import static common.CommonActions.*;
+
+
```
Comment (Goldbe54) [inline src/main/java/fragments/FilterFragment.java:8]: delete empty line

**Review by Goldbe54 — APPROVED** [review]

**Pull request merged**